extensions/2.0/Khronos/KHR_displaymapping_pq/README.md

```diff
@@ -0,0 +1,439 @@
+According to research (Bernstein et al. 2018) the mind's perception of brightness is mostly the same, regardless of object brightness.  
+The difference in the perception of the object's brightness is based on background color.  
+This means that in order to have deterministic control of perceived brightness it is important to be able to control both object and background color, for instance by means of a set dynamic range.  
+
+It also provides the specification for using HDR compatible display outputs while at the same time retaining compatibility with SDR display outputs.  
+
+The intended usecases for this extension is any usecase where the light contribution values will go above 1.0, for instance by using KHR_lights_punctual, KHR_emissive_strength or KHR_environment_lights.  
+When using this extension it is recommended that the dynamic range is kept between 0 and 10 000 lumen/m2 in order to utilize the range of the Perceptual Quantizer.  
+This can for instance be done by exporters by scaling light contributions values.  
+
+Here the term rasterizer means a rendering engine that consits of a system wherein a buffer containing the pixel values for each frame is prepared. 
+This buffer will be referred to as the framebuffer.  
+The framebuffer can be of varying range, precision and colorspace. This has an impact on the color gamut that can be displayed.  
+
+After completion of one framebuffer it is output to the display, this is usually done by means of a swap-chain. The details of how the swap works is outside the scope of this extension.  
+KHR_displaymapping_pq specifies one method of mapping internal pixel values to that of the framebuffer.  
+
+This extension does not take the viewing environment of the display, or eye light adaptation, into consideration.  
+It is assumed that the content is viewed in an environment that is dimly lit (~5 cd / m2) without direct light on the display.  
+Viewer calibration is not part of this extension as this is heavily dependant on the usecase and application.  
+
+[Output using extension, without scene aperture, and one directional light with an intensity of 10 000 lumen / m2](images/DisplayMappingPQ-10000.png?display=inline-block)
+
+
+
+
+
+### glTF asset considerations
+
+The extension affects the entire glTF asset, ie all scenes, geometry, images and textures, included in a file that is using this extension.   
+This means that the current rendered scene shall be output using the displaymapping declared by this extension whenever the usecase is relevant, ie a realtime rasterizer with output to a display.  
+
+Visualization of multiple glTF assets using this extension is supported and will produce a normative result.  
+
+If the glTF asset contains multiple scenes, each one when rendered, shall be output using this extension.  
+
+If the glTF asset contains this extension but no scene or model data then it may be treated as an enabler for displaymapping.  
```

Comment:
Ok, but section 3.5 of the spec states:
```
A glTF asset that does not contain any scenes SHOULD be treated as a library of individual entities
such as materials or meshes.
```
Would you prefer that the behavior of using KHR_displaymapping_pq but not containing any scenes is undefined?

Reply:
Since the aperture controls could be defined per-scene, I would suggest one of these two options:

### 1. Unify the controls
- Apply the PQ only to _scenes_ that have the `KHR_displaymapping_pq` extension object. Aperture controls may still be optional as they are now.
- Note that meshes not attached to nodes and nodes not attached to scenes are not rendered anyway.

### 2. Further split the controls
- Do not define any new objects for the `KHR_displaymapping_pq` extension, leave it only as an element of the `extensionsUsed` array.
- When the extension is present, it enables PQ output for all scenes of the asset.
- Note that meshes not attached to nodes and nodes not attached to scenes are not rendered anyway.
- Move aperture controls to a new extension that would require this one.

Reply:
Hi @lexaknyazev and thanks for your comments and suggestions!

Looking at this extension from a 3D Commerce (and ikea) point of view:
What we want to achieve is to mark an asset to be rendered using this extension (retaining hue) and also be confident that a well behaved renderer (certified) will handle the asset accordingly under different usecases.

If we choose number 1:
* What would happen to an asset (nodes) that is loaded into an existing scene that does not use the extension?

If we choose number 2:
* I am thinking about going this route - the problem is that this may lead to situations where the hue shift is still present simply because users have lights above 10 000.
Do you see a solution for this scenario?
Is it as simple as specifying that light contribution values when rendering this asset must be within [0 - 10000] or do you forsee other problems?

Thanks for your constructive comments @lexaknyazev

Reply:
> What would happen to an asset (nodes) that is loaded into an existing scene that does not use the extension?

- Loading multiple glTF assets into an existing rendering environment is generally not defined by the core spec.
- In my opinion, it's quite unexpected for the newly added asset to change the rendering output of pre-existing assets. Should the global PQ be turned off when the asset is removed later?
- For the option 1, I'd just discourage applications from mixing PQ and non-PQ assets. Bulk-adding this extension string to the existing library is trivial enough.

>  the problem is that this may lead to situations where the hue shift is still present simply because users have lights above 10 000

This problem remains anyway because the current extension spec allows omitting the aperture controls and says:
> Light contribution values above 10000 cd / m2 are discouraged and will be clamped by implementations if not adjusted by scene aperture.

Reply:
Good points!

>This problem remains anyway because the current extension spec allows omitting the aperture controls and says:

Correct - and I feel that this is the part of this extension that is not decided on.
The more I think about it the more I feel the need for fixed aperture control, ie option 2 with fixed values of:

value = 10 000 
minLuminance = 0
maxLuminance = naively collected on a per frame basis from light contribution. If total light > 10 000 then all contribution will be scaled according to highest value.
control = removed

Do you think this would help with the purpose of this extension?

Reply:
@lexaknyazev It would be super great if you are available for a quick call on teams?
(rickard.sahlin at inter.ikea.com)

Reply:
I do think that exposure and tone mapping is more of a camera property than a glTF file property.  Also putting it on the camera makes it clear what to do when combining multiple GLTF files together in the same scene.

Reply:
Hi @bhouston and thanks for your comment

Yes, I also think so.
Currently the aperture setting is set to a default behavior that will do nothing if values are within [0 - 10000] range - but if lit with high intensity light > 10000 then the aperture function will scale incoming light so that the highest value is 10000.
Ie meaning that if you have a light with intensity 10000 and add a new light with intensity 100000 - then they will become 10000 and 1000 respectively.

Also, the extension is turned on simply using the KHR_displaymapping_pq string in the asset.
All in order to simplify and make the extension work the same regardless of glTF model/scene.

Do you think this will suffice?